WIP - Do Not Merge - BackgroundColorBehavior Layer fix

podraco opened the pull request:
Fix for specific background behavior.

This fix ensures that the background layer is drawn before the active 
content of the widget..

### Description of the problem
At some point during development, the background color layer was moved from the bottom layer to the middle layer, where most widgets with texture draw their data, since this behavior is stackable, it will be drawn after the previous behavior, in the case of the MDLabel, the background layer is drawn over the text texture layer.

### Reproducing the problem

```python
from kivy.lang.builder import Builder
from kivymd.app import MDApp


kv="""
FloatLayout:
    MDLabel:
        halign: "center"
        valign: "center"
        md_bg_color: (1,0,0,0.5)
        size: 120,64
        size_hint: None,None
        pos_hint: {"center": [0.5, 0.5]}
        text: "this is visible text"
        text_size: self.size
"""

class app(MDApp):
    def build(self, *dt):
        return Builder.load_string(kv)

app().run()
```

### Screenshots of the problem
![image](https://user-images.githubusercontent.com/5509325/135308877-5e69a92b-7212-424d-8166-ad217cb33d3d.png)
here, we can se that the background color is drawn over the text layer, which means that the text will be hidden behind the texture if the alpha channel is 1.

### Description of Changes
changed the layer where the background color is drawn, moved from canvas to canvas,before.
Describe what changes you create to solve the problem.

### Screenshots of the solution to the problem
![image](https://user-images.githubusercontent.com/5509325/135309319-cb6f8847-42c0-4f44-8f9d-da76246d060e.png)
in this image, we set the alpha channel of the md_bg_color to 1 instead of the 0.5 that is on the example code.

Attach screenshots that demonstrate that your changes solved the problem.

podraco committed:
BackgroundColorBehavior Layer fix

Fix for specific background behavior.

This fix ensures that the background layer is drawn before the active 
content of the widget..

since at some point it was moved to the front layer, which collided and overlapped with some other widget's textures, such as labels and images.

HeaTTheatR commented:
@podraco Let me check it out tomorrow...

HeaTTheatR commented:
@podraco
Your changes break the display of other widgets:

![Снимок экрана 2021-09-30 в 14 20 18](https://user-images.githubusercontent.com/16930280/135446176-89a119c2-9c66-436b-a925-856397e953f2.png)

HeaTTheatR commented:
@podraco 
![Снимок экрана 2021-09-30 в 14 22 13](https://user-images.githubusercontent.com/16930280/135446381-19769c62-9376-4b30-9e48-a1e043a48877.png)

podraco committed:
Fix layer issue with rounded buttons

Fixed missplacement of the background layer on the following objects.
BaseRoundButton
BaseButton
this fix keeps every button the same way, drawing the background layer on the canvas.before layer instead of the middle, as the rest of the classes do.

HeaTTheatR commented on kivymd/uix/button/button.kv:
```diff
@@ -32,7 +32,7 @@
 
 <BaseRoundButton>
-    canvas:
+    canvas.before:
```
Your changes break the elevation effect in the buttons.

HeaTTheatR commented on kivymd/uix/button/button.kv:
```diff
@@ -1,5 +1,5 @@
 <BaseButton>
-    canvas:
+    canvas.before:
```
Your changes break the elevation effect in the buttons.

podraco commented:
> @podraco Your changes break the display of other widgets:
> 
> ![Снимок экрана 2021-09-30 в 14 20 18](https://user-images.githubusercontent.com/16930280/135446176-89a119c2-9c66-436b-a925-856397e953f2.png)

if you read the buttons kv code, you can see that the code is broken for those, they should be drawing inside the "before" layer of the cnavas, not the "current."

all tho i'm not sure how the canvas change broke the chips, i don't see them working before or after the change, it might be something else, as its the exact same image you sent me.
this is done from the master.zip file.
![image](https://user-images.githubusercontent.com/5509325/135499694-c173abe6-ecf9-4a12-8936-cdb04ce7b62b.png)

HeaTTheatR commented:
@podraco 
![Снимок экрана 2021-09-30 в 20 04 06](https://user-images.githubusercontent.com/16930280/135499832-0cbd5068-259d-4e16-9a2e-04aa13173823.png)

podraco commented:
> @podraco ![Снимок экрана 2021-09-30 в 20 04 06](https://user-images.githubusercontent.com/16930280/135499832-0cbd5068-259d-4e16-9a2e-04aa13173823.png)

true i forgot there is a clear instruction.

podraco commented:
ok i'm doing some deep research and so far  i can see that the color behavior is not respected across the app, instead, there are multiple individual instances of the same "standard" behavior. so i'm updating most of the widgets i see with the issue.

## This merge request will be set to WIP on the meantime, 

As I debug and check, whatever breaks.

This merge request will only concern about the color management behavior.

podraco committed:
Added new PreCreated Color behavior objects.

These object will allow us to add the same colorbehavior base class with 
the added canvas instruction, this will help us to normalize the 
behavior across the library.

Added new color behaviors to the __init__ file on kivymd behaviors

podraco committed:
replaced base name root/self.md_bg_color with *._md_bg_color

this change will allow users to easely change the color of the widgets 
wihtout overwritting the previos one, and thus, make it easier to modify 
the status of the widgets by useing the normal state of the button.

for example, if you get a widget in a group, you can set the current 
color to the accent colour of the theme and when it's not active, just 
update it's color to the base color that is the object.md_bg_color.

HeaTTheatR commented:
@podraco It's a good news!

podraco committed:
Added the meta behaviorclass Background Color Behavior to basebutton

this will allow us to easely change properties in each button the same 
way always.

simplified some logic with the use of the _md_bg_color property, this 
will make easier for developres to toggle between origings, either by 
setting the theme color or by allowing the use of a custom color, for 
example

The meta class will always start with
`obj.md_bg_color=None`
which means that we are now able to do some simple logic as the 
following:

`obj._md_bg_color: self.md_bg_color if self.md_bg_color else 
app.theme_cls.primary_color`

instead of comparing the value to an existent colour. which can lead to 
some positive false errors.

this update normalizes the buttons with the expected programming 
standard provided by the color behavior class.

podraco committed:
Color behavior update:

### MDToolbar

Replaced set_md_bg_color with update_bg_color, this is done because 
update_bg_color is provided already by the color behavior.

which will ensure the correc theming display. as well will diminish the 
ammount of events dispatched.

removed set_md_bg_color from MDToolbar

### MDBottomAppBar
added BackgroundColorBehavior to MDBottomAppBar class definition.

### NotchedBox :: KV file
Simplified color management event logic. instead of doing 3 
comparations, it has been lowered to 1, this logic won't interfere with 
the python side and will allow the development team to set the proper 
theme color on the object.update_bg_color method.

podraco committed:
added typing
bugfix, solved the issue in which the update color is dispatched with no 
current md_bg_color assigned.

podraco committed:
COLOR UPDATE :: ADDED UPDATE_BG_COLOR TO NOTCHEDBOX.

*Added typing

*Added update_bg_color method to NotchedBox, this method will manage the 
theming updates properly, if there is no md_bg_color, the current color 
will be set to the primary color. of the theme.

podraco committed:
COLOR UPDATE :: MDCARD :: Bound background color to _md_bg_color

this update exchanges the color management to the new standard mehtor 
that is provided by the color behavior update.

podraco committed:
COLOR UPDATE :: MDCHIP :: canvas instructions moved to canvas.before

* Added clear instruction to remove any unnecesary instruction.

podraco committed:
COLOR UPDATE :: TOOLBAR :: Added missing event

this event management will set the background color to the theme's 
primary color if there is no md_bg_color assigned.

podraco committed:
COLOR UPDATE :: SNACKBAR :: rebounded background color to new standard

bound Color to _md_bg_color instead of md_bg_color. this is done in 
favor of the new standard provided by the color behavior.

podraco committed:
COLOR UPDATE :: Navigationrail :: rebounded background color

bound Color to _md_bg_color instead of md_bg_color. this is done in 
favor of the new standard provided by the color behavior.

podraco committed:
COLOR UPDATE :: MDNavigationDrawer :: rebounded background color

bound Color to _md_bg_color instead of md_bg_color. this is done in 
favor of the new standard provided by the color behavior.

podraco committed:
COLOR BEHAVIOR :: BUGFIX

Added event dispatchement on the first frame abailable, this will ensure 
that the color is displayed properly.

HeaTTheatR commented:
@podraco What changes have occurred? What do your new commits fix/add? Or is this PR still in development?

podraco commented:
this PR is still in development, i'm still checking that the widgets are not broken
so far i removed canvas instructions from the bacground color behavior as most of the widgets in the lib use the Clear insturction to remove them, i'm thinking of letting a small call that adds the instructions after the instance is created that is by deffault True.
what do you think?

HeaTTheatR commented:
@podraco If this PR fixes the current problems, then that's a good thing.

podraco commented:
this fixes every widget because the color behavior draws over the current layer, and widgets such as buttons, text and others, are affected by this.
specially MDLabel and MDIcon are affected by this, hiding the text texture below the background texture.

podraco commented:
it's still a work in progress

podraco committed:
Fixes theming issue, front layer of the backdrop widget color issue.

the front layer color was removed, for some unkwonwn reason. as the 
theme was changed.

podraco committed:
COLOR UPDATE :: MDBACKDROP :: fixed color and canvas issues.

* The back layer has been resized to fit correctly the provided space 
behind the front layer. it's also aligned with the top property.

* some documentation has been added to allow developers understand how 
do the MDBackDrop widget works.
* Added documentation for the elevation property, this one will be 
parsed to the front layer, since it requires a Fake elevation behavior 
to display anything.
* The backdrop now works inside a Stencil view, this has been done to 
improve performance, as it will clip anything that is not needed and 
it's outside the render view.
* Improved theming options
* back_layer_color is now deprecated and will be removed soon. this has 
been done in pos of the use of the md_bg_property, that already exists 
and has all the theming logic bound.
* The toolbar background color is now bound to root._md_bg_color
* Backlayer is now a Grid Layout instead of the box layout that was 
previously, this changed because the Grid layout already fits the 
content from top to bottom and from left to right.

podraco committed:
BUGFIX :: TIME INPUT TEXT FIELD  :: TIME PICKER COMPONNENT.

this fix removed a bound method that was leading to a long removed 
property from the md text input widget.

podraco committed:
BUGFIX :: TimeInputTextField :: TIME PICKER COMPONENT.

Bound `disabled_foreground_color` to `text_color`.
this is a hot fix, since the widget uses the disabled status as a 
display state, rather than adding an actual new state for the widget.
this might need some further improvement.

podraco committed:
COLOR UPDATE:: MDBOXLAYOUT :: Added canvas.before to MDBoxLayout.

* The instruction added is a Rectangle canvas instructions. 
* Color is bound to self._md_bg_color

podraco committed:
Added custom classes to the __all__ memory list.

podraco committed:
COLOR UPDATE :: KITCHEN SINK :: fixed layout issues on start screen.

FIXED MD BACKDROP WIDGET LAYOUT.
this fixes the issue where the Float layout deppended on the wrong size, 
as it was calculated upon the top box and was bigger than the actual 
container.

Fixed MD checkbox issue: now the check boxes will display correctly what 
theme is selected and the correct one will keep selected upon theme 
changes.

podraco committed:
COLOR UPDATE :: MDFLOATLAYOUT :: Added Canvas.before to MDFloatLayout.

added Rectangle as the texture instruction and the color is bound to 
_md_bg_color

podraco committed:
added Rectangle as the texture instruction and the color is bound to 
_md_bg_color

podraco committed:
BUGFIX :: MD LAYOUTS

These layouts has the wrong name for the builder method call, this 
commit fixes them.

podraco committed:
COLOR UPDATE :: MDScreen :: Added Canvas.before to MDScreen.

added Rectangle as the texture instruction and the color is bound to 
_md_bg_color

podraco committed:
COLOR UPDATE :: MDRelativeLayout :: Added Canvas.before.

added Rectangle as the texture instruction and the color is bound to 
_md_bg_color

podraco committed:
COLOR UPDATE :: MDStackLayout :: Added Canvas.before.

added Rectangle as the texture instruction and the color is bound to 
_md_bg_color

podraco committed:
COLOR UPDATE :: MDLABEL :: Added label canvas.before instructions.

* Color is bound to _md_bg_color
* Vertex Instruction: RoundedRectangle
* Allows radius assignment.

podraco commented:
Ok, so far, the current Color management issue is now fixed, I reviewed all the Kitchen sink demo and i didn't find anything out of the usual.
The same issues that were present and were not about the color management are still present, except for the backdrop layout, i updated it, so it would show the canvas properly. 

For some reason, the canvas instruction from the front layer color was set to [0,0,0,0] whenever the theme changed, i'm not sure why.

Besides that, it looks ready to be merged, the current conflict is about the same commit we did, it removed the binding method to a deprecated and removed property.

i would appreciate if you could review this, so we can polish the remaining details.

podraco commented:
Ok, so far, the current Color management issue is now fixed, I reviewed all the Kitchen sink demo and i didn't find anything out of the usual.
The same issues that were present and were not about the color management are still present, except for the backdrop layout, i updated it, so it would show the canvas properly. 

For some reason, the canvas instruction from the front layer color was set to [0,0,0,0] whenever the theme changed, i'm not sure why.

Besides that, it looks ready to be merged, the current conflict is about the same commit we did, it removed the binding method to a deprecated and removed property.

i would appreciate if you could review this, so we can polish the remaining details.

HeaTTheatR commented:
@podraco 

https://user-images.githubusercontent.com/16930280/135929172-cbf3d06b-7918-4175-b5ff-16948b0b2c5c.mov

HeaTTheatR commented:
@podraco 

This PR:

![Снимок экрана 2021-10-05 в 00 28 03](https://user-images.githubusercontent.com/16930280/135929238-c09ec73f-5a9f-48a3-ab66-f416869895ef.png)

Current state:

![Снимок экрана 2021-10-05 в 00 27 43](https://user-images.githubusercontent.com/16930280/135929286-678e7c71-62c3-4f18-9b20-c6ffe47de2b1.png)

HeaTTheatR commented:
@podraco 

![Снимок экрана 2021-10-05 в 00 29 02](https://user-images.githubusercontent.com/16930280/135929332-12ea9008-0a4a-4410-b6ad-6107062c6769.png)

HeaTTheatR commented:
@podraco 

![Снимок экрана 2021-10-05 в 00 49 42](https://user-images.githubusercontent.com/16930280/135929627-35854f9e-db4f-405d-9bb9-d470a2c781b7.png)

HeaTTheatR commented on demos/kitchen_sink/libs/kv/start_screen.kv:
```diff
@@ -127,7 +134,8 @@ MDBoxLayout:
                 icon: "palette-outline"
                 divider: None
-                on_release: app.show_dialog_change_theme()
+                on_release:
+                    app.show_dialog_change_theme()
```
Use line breaks only if the line is longer than 79 characters.

podraco replied:
it's an event, event's must be line broken, since they are logic blocks, this makes it easier to identify them without having to read the whole configuration. Specially after a while reading code.

HeaTTheatR commented:
@podraco 
Also, since your local changes are older, I was unable to test the MDDataTable widget with new row color features:

```python
from kivy.metrics import dp
from kivy.uix.anchorlayout import AnchorLayout
from kivy.utils import get_color_from_hex

from kivymd.app import MDApp
from kivymd.uix.datatables import MDDataTable


class Example(MDApp):
    def build(self):
        layout = AnchorLayout()
        data_tables = MDDataTable(
            background_color_header=get_color_from_hex("#65275d"),
            background_color_cell=get_color_from_hex("#451938"),
            background_color_selected_cell=get_color_from_hex("e4514f"),
            size_hint=(0.9, 0.6),
            use_pagination=True,
            column_data=[
                ("No.", dp(30)),
                ("Column 1", dp(30)),
                ("Column 2", dp(30)),
                ("Column 3", dp(30)),
                ("Column 4", dp(30)),
                ("Column 5", dp(30)),
            ],
            row_data=[
                (f"{i + 1}", "1", "2", "3", "4", "5") for i in range(50)
            ],
        )
        layout.add_widget(data_tables)
        return layout


Example().run()
```

podraco commented:
As an update on this Merge request, I created a branch on my repo, that i'm working right now, i'm checking the buttons so it doesn't create multiple canvas instructions and added some color functions.

![image](https://user-images.githubusercontent.com/5509325/136474735-b92452d1-d532-4bc5-ae69-e6d51c6f93f2.png)
In this image, you can see a grid of buttons that maps only the color definition palete (by column) and their wheights (by row), the square in the center is the new mehtod that process it's contrast and decides either white or black text is required.

* added a method that can compute the proper text color based on the YIQ color space of a defined color. 

* And added 2 new palettes to the color definition:  `"background_dark", "background_light"`; this aims to increase contrast between the current theming.bg_* properties. you can see them in the image, the last 2 columns of hte image.

* Added some logic to allow buttons to auto-calculate their `text_color` or use the developer's provided one (front end property: `obj.text_color`, backend one: `obj._text_color`) these methods are stored inside the Theming class.

* Made that the call to self.theme_cls goes thought a proxy instead of creating a hard link to the app's property, works exactly as before, but it's lighter, since now, it's just a "pointer" to that object and can avoid some gc errors and memory leaks.

* The bg_dark, bg_light, bg_normal and bg_darker are now linked to the A700,A400, A200, A100 weights on the new color palettes added.

* So far, made somewhat lighter the MDFLatButton and some basebutton classes, just keeping the necessary data management to each class and adding the canvas to the end button or the base_behavior class that is shared. so, deleting any canvas instruction that is not required or configurable by the button class.

i'm still doing some cleanup and testing.
So far, the MDFlatButton and base button doesn't have the issue where if you changed theme, the custom values were deleted and replaced by the theme colors.

once i have that branch ready, i'm gonna merge it with my master and continue with the next classes that require a review. this way it's easier to maintain and revert anything that doesn't work properly.

HeaTTheatR commented:
> As an update on this Merge request, I created a branch on my repo, that i'm working right now, i'm checking the buttons so it doesn't create multiple canvas instructions and added some color functions.
> 
> ![image](https://user-images.githubusercontent.com/5509325/136474735-b92452d1-d532-4bc5-ae69-e6d51c6f93f2.png) In this image, you can see a grid of buttons that maps only the color definition palete (by column) and their wheights (by row), the square in the center is the new mehtod that process it's contrast and decides either white or black text is required.
> 
> * added a method that can compute the proper text color based on the YIQ color space of a defined color.
> * And added 2 new palettes to the color definition:  `"background_dark", "background_light"`; this aims to increase contrast between the current theming.bg_* properties. you can see them in the image, the last 2 columns of hte image.
> * Added some logic to allow buttons to auto-calculate their `text_color` or use the developer's provided one (front end property: `obj.text_color`, backend one: `obj._text_color`) these methods are stored inside the Theming class.
> * Made that the call to self.theme_cls goes thought a proxy instead of creating a hard link to the app's property, works exactly as before, but it's lighter, since now, it's just a "pointer" to that object and can avoid some gc errors and memory leaks.
> * The bg_dark, bg_light, bg_normal and bg_darker are now linked to the A700,A400, A200, A100 weights on the new color palettes added.
> * So far, made somewhat lighter the MDFLatButton and some basebutton classes, just keeping the necessary data management to each class and adding the canvas to the end button or the base_behavior class that is shared. so, deleting any canvas instruction that is not required or configurable by the button class.
> 
> i'm still doing some cleanup and testing. So far, the MDFlatButton and base button doesn't have the issue where if you changed theme, the custom values were deleted and replaced by the theme colors.
> 
> once i have that branch ready, i'm gonna merge it with my master and continue with the next classes that require a review. this way it's easier to maintain and revert anything that doesn't work properly.

It sounds good! Please note the changes in the `MDBottomNavigation` class:

https://github.com/kivymd/KivyMD/commit/57f1aa9232e236bbd4eb658eb20273ce4cc06c3b
https://github.com/kivymd/KivyMD/commit/3aa9ad7004a2199bd1c59f9f0f17c039776e86ed

This is the `kitchen_sink` branch. I shouldn't have made these commits as I'm working on a new demo app on this branch and shouldn't have touched library modules directly...

podraco commented:
i'll merge them with the branch latter today or tomorrow.

HeaTTheatR commented:
> i'll merge them with the branch latter today or tomorrow.

Also in this branch the demo application has been removed.

podraco commented:
as a heads up, the pull request i'm making for kivy's master branch, allows us to do the following:

![UJ0sWTXa1C](https://user-images.githubusercontent.com/5509325/137223281-1946809c-1b92-4967-b13d-c5db7092e048.gif)
both are rounded buttons, jusct changed the segment value to 1,2 or 12 for rounded corners.
still workikng on some issues i found.

podraco commented:
This merge request is meant for kivy's master after https://github.com/kivy/kivy/pull/7660 is merged within the master branch.

This means that It might display bugs in older versions of kivy. Because of how kivy's vertex_instruction_line managed the rounded_rectangle line instruction before.

HeaTTheatR commented:
@podraco We can mark the minimal version of Kivy in the dependencies.

HeaTTheatR commented:
Deprecated
